### ISSUE TITLE
Server rejects certain sizes of streamed request bodies

### ISSUE THREAD

**Issue by Mahoney** [eclipse/jetty.project#9990]
**Jetty version(s)**
11.0.15

**Java version/vendor** `(use: java -version)`
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7)
OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode)

**OS type/version**
macOs 13.4.1 (22F82)

**Description**
When posting certain body sizes to a JAX-RS server via a Jersey client, the server returns a 500 because of an exception thrown inside Jetty:

```
127.0.0.1 - - [28/Jun/2023:16:49:19 +0000] "POST /test HTTP/1.1" 500 110 "-" "gzip-interceptor-bug (RQZNuukVWd)" 1
ERROR [2023-06-28 16:49:19,766] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: 20dae4a5c6058e23
! java.io.IOException: Interceptor org.eclipse.jetty.server.handler.gzip.GzipHttpInputInterceptor@1aa1d204 did not consume any of the 8 remaining byte(s) of content
! at org.eclipse.jetty.server.AsyncContentProducer.intercept(AsyncContentProducer.java:449)
! at org.eclipse.jetty.server.AsyncContentProducer.transformRawContent(AsyncContentProducer.java:378)
! at org.eclipse.jetty.server.AsyncContentProducer.nextTransformedContent(AsyncContentProducer.java:368)
! at org.eclipse.jetty.server.AsyncContentProducer.nextContent(AsyncContentProducer.java:250)
! at org.eclipse.jetty.server.BlockingContentProducer.nextContent(BlockingContentProducer.java:103)
! at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:287)
! at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:272)
! at org.glassfish.jersey.message.internal.EntityInputStream.read(EntityInputStream.java:79)
! at java.base/java.io.InputStream.readNBytes(InputStream.java:409)
! at java.base/java.io.InputStream.readAllBytes(InputStream.java:346)
! at foo.TestResource.absorbBytes(TestResource.java:20)
```

**How to reproduce?**

Check out this repo and run `mvn test`:
https://github.com/Mahoney-bug-examples/gzip-interceptor-bug

**Further context**

This bug was not present in Jetty 9.4.50.v20221201 (Dropwizard 2.1.7)

**Comment by Mahoney** [comment]
I raised this against the Dropwizard project in https://github.com/dropwizard/dropwizard/issues/7275 , but the more I think about it the more likely it seems to be a Jetty bug. Apologies if it's bad form to raise against both.

**Comment by Mahoney** [comment]
Resource:

```java
@Path("/test")
public class TestResource {

    @POST
    @Consumes(MediaType.APPLICATION_OCTET_STREAM)
    @Produces(MediaType.APPLICATION_JSON)
    public Response absorbBytes(InputStream bytes) throws IOException {
        return Response.ok(Map.of("bytes_received", bytes.readAllBytes().length)).build();
    }
}
```
https://github.com/Mahoney-bug-examples/gzip-interceptor-bug/blob/main/src/main/java/foo/TestResource.java

Test:
```java
@ExtendWith(DropwizardExtensionsSupport.class)
public class RequestSizeTest {

    private static final DropwizardAppExtension<BugConfiguration> APP = new DropwizardAppExtension<>(
            BugApplication.class
    );

    @ParameterizedTest
    @ValueSource(ints = {
            8191,
            8192,
            8193,
            8194,
            8195,
            // many others, redacted for length...
    })
    void can_post_arbitrary_body_sizes_to_the_server(int bytesToSend) throws Exception {

        var body = new StreamingOutput() {
            @Override
            public void write(OutputStream output) throws IOException, WebApplicationException {
                for (int i = 0; i < bytesToSend; i++) {
                    output.write(1);
                }
            }
        };

        try (Response response = new JerseyClientBuilder(APP.getEnvironment())
                .build(RandomStringUtils.randomAlphabetic(10))
                .target("http://localhost:" + APP.getLocalPort())
                .path("/test")
                .request()
                .post(Entity.entity(body, APPLICATION_OCTET_STREAM))) {

            var data = response.readEntity(String.class);
            assert response.getStatus() == 200 : data;
            assertEquals(Map.of("bytes_received", bytesToSend), new ObjectMapper().readValue(data, Map.class));
        }
    }
}
```
https://github.com/Mahoney-bug-examples/gzip-interceptor-bug/blob/main/src/test/java/foo/RequestSizeTest.java

**Comment by joakime** [comment]
Confirming bug.

I was able to reproduce in a test case in the `jetty-10.0.x` HEAD.

See (failing) testcase in branch `fix/10.0.x/gzip-input-decompress-bug`.

* https://github.com/eclipse/jetty.project/blob/fix/10.0.x/gzip-input-decompress-bug/jetty-servlet/src/test/java/org/eclipse/jetty/servlet/GzipHandlerInputTest.java)

My first thought was that this is a `Transfer-Encoding: chunked` side effect, but this happens in both cases

1. Client uses `Transfer-Encoding: chunked`
2. Client uses `Content-Length: ###`

Now to dig into what is happening.

**Comment by joakime** [comment]
Opened PR #9997 to address this

**Comment by joakime** [comment]
Turns out we had some old code to prevent infinite loops from interceptors that make no progress that is getting in the way for this issue.
That infinite loop prevention code was removed in Jetty 12, so we essentially reproduced that effort down here in Jetty 10.0.x as well. (will be merged to 11.0.x after PR approval/merge)

**Comment by gregw** [comment]
@joakime has this been merged forward to 12?

**Comment by joakime** [comment]
Merged PR #9997 into `jetty-10.0.x` and merged up to `jetty-11.0.x`
